common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -353,22 +355,22 @@ protected void reportInstancesLeak(String resourceType) {
         @SuppressWarnings("unused")
         private volatile int droppedRecords;
 
-        private final ConcurrentMap<DefaultResourceLeak<?>, LeakEntry> allLeaks;
+        private final Set<DefaultResourceLeak<?>> allLeaks;
         private final int trackedHash;
 
         DefaultResourceLeak(
                 Object referent,
                 ReferenceQueue<Object> refQueue,
-                ConcurrentMap<DefaultResourceLeak<?>, LeakEntry> allLeaks) {
+                Set<DefaultResourceLeak<?>> allLeaks) {
             super(referent, refQueue);
 
             assert referent != null;
 
             // Store the hash of the tracked object to later assert it in the close(...) method.
+            trackedHash = System.identityHashCode(referent);
```

Comment:
please move this where it was before as I think the rest of the comments also belongs to it. It explains why we can not just store referent itself.